### DISCUSSION TITLE
October 2023 Workshop Release Notes

### DISCUSSION THREAD

**Discussion by Denneisk** [wiremod/wire#2828] [General]
*Update 2023-11-09: Delta operator was reverted back to its original functionality*

Here's the October 2023 Wire release notes! Hopefully one of many!

There's been lots of changes with E2 this month, and we're excited to finally release them to the Workshop.

## The E2 Compiler Rewrite
The E2 compiler has been rewritten. Vurv has been working on this for quite some time on this (since December!) and we're proud to finally release it to the Workshop.
Extension developers, __your extensions may need to be updated to fit the changes!__ [See a full list of updates to the API here.](https://github.com/wiremod/wire/pull/2555)
Despite being a huge change for E2 internally, for users, the E2 compiler rewrite should be mostly unnoticeable. Some user-facing improvements include improved errors and warnings, but more importantly, your chips should use less ops and CPU time! Performance has been a huge focus on the compiler rewrite and we're hoping you won't be disappointed.
And, please pardon us if you notice a bug. We've worked hard trying to stomp out anything that's come up during this month, but if you happen to encounter one, [please let us know](https://github.com/wiremod/wire/issues)!

### Noteworthy functional differences:
- For user-defined functions, function returns are now *required* for non-void functions. The compiler will error if your code does not. Additionally, look forward to [compile-time functions](https://github.com/wiremod/wire/pull/2789), which will bring some major changes to functions in E2.
- The `$` operator now operates slightly differently, and instead now changes its value each time it's used. [This may be getting fixed to previous behavior](https://github.com/wiremod/wire/pull/2794), but the `$` operator is going to be deprecated. To future-proof your code, please use an alternative to `$`, such as using a persistent variable.
- The operation of arrays with `foreach` has changed slightly. Previously, you were able to use `foreach` over a non-sequential array, but this has been changed so that `foreach` now only works on sequential array keys, starting from 1. To get non-sequential access, use a table instead.
- For complex numbers or quaternions, the constants `i`, `j`, and `k` are now prefixed with a 1, as `1i`, `1j`, and `1k`.

## E2 Additions
And if you thought the E2 compiler being rewritten wasn't enough, we've got plenty of new additions to the E2 library that are likely to be useful to you.

First off, we have [a new E2 event, `playerMove`](https://github.com/wiremod/wire/pull/2779), which brings along two types with it. You can expect this to work like a more accurate `keyPressed` event except focused on precise player movement. Especially, users may like the `usercmd:getMouseDeltaX()` and `usercmd:getMouseDeltaY()` functions, which get precise mouse movement data.

Next, we have some pretty big changes to *two* E2 libraries. First off, the [*file* extension has had a much-needed update](https://github.com/wiremod/wire/pull/2763) to fix some glaring bugs and some modernized features. Files can now be loaded up to 5 (by default) queued operations. This change will make multi-file projects a *lot* easier to work with. Additionally, *file* has finally received a proper *eventification* update, so if you haven't made the switch already, now is the perfect time. And on top of all that, a new event just for writing files, `fileWritten`, has been added, that triggers when a file finishes writing.

After that, we have a huge change to how you can create EGP objects. As part of the modernization of Wiremod, we've been working on removing index-based systems in favor of direct object references. In less technical terms, we're releasing the *egpobject* type along with the *egpobjects* extension, providing a new and __superior__ method for interacting with EGPs. For more information, and some examples of the completely new things you can do, check out [the PR details here](https://github.com/wiremod/wire/pull/2646).

Finally, as part of our efforts to decrease reliance on third-party E2 extensions, we're releasing a little addition that may be of use to at least a few people; you can now [adjust flexes and eye targets](https://github.com/wiremod/wire/pull/2810) on models!

That's all for this month's progress! We hope you enjoyed reading this and that you will enjoy these changes! Let us know what you think about these newsletters. In the meantime, this is where I get off.

# Full changelog:

### Additions
- Added *Entered* output to Text Entry (*Community Contribution*)
- Reduced Friends List maximum size from infinite to 2048
### Fixes
- Fixed Interactive Prop sending invalid entities in some cases
- Fixed Wire screens would show incorrect values at certain precisions
- Fixed a memory leak with Text Entry (*Community Contribution*)

### Additions
- Added more compile-time errors and warnings
- Add `event playerMove` and types `usercmd` and `movedata` for getting more detailed player movement information (*Community contribution*)
- Add `setOverlayText(string)` to change the overlay text of individual E2s
- Add `canSetName` to see if chip can use `setName`
- Added helper functions `file<Load/Write/List>Queued` to get the current queued amount
- Added `event fileList` as event alternative to `fileClk`
- Added `event fileWritten` that executes when a file finishes sending to the client
- [Added `egpobject` type and *egpobjects* extension](https://github.com/wiremod/wire/pull/2646). Full list of functions can be found in the E2 Helper under the `egpobjects` extension.
- Added `entity:setFlexWeight` and `:setFlexScale` to manipulate model flexes (face posing), as well as various getters related to them
- Added `entity:setEyeTarget` and two similar functions to allow model eye posing
### Changes
- [Rewrote the E2 compiler](https://github.com/wiremod/wire/pull/2555)
- Improved some compile-time errors and warnings
- Improved net code for hologram clips
- Decreased various op costs
- Allow *constraintcore* to constrain to the world entity
- Enforced non-void user-defined functions to return
- Modified restrictions on `setName` to be more lenient
- Modified *file* extension to allow users to queue up to 5 (default) files for sending/receiving
- Improved outputs and errors on various *file* functions
- Modified op costs on file functions to better reflect their performance
- Improved interals on EGP to be easier to work with and more efficient
- Changed `entity:bones()` to return a copy of the bones table to avoid exploitation
### Fixes
- Fixed switch statements evaluating their operators multiple times
- Fixed an issue with complex constants and function signatures overlapping
- Fixed hologram clips could underflow
- Fixed `entity:getBodyGroups` accidentally erroring (*Community contribution*)
- Fixed programmatically generated functions being placed in the wrong extension
- Fixed EGP poly position breaking on some E2 functions
- Fix returning from functions inside switch statements didn't execute properly
- Fixed an exploit with effects that let users bypass the filter (*Community contribution*)
- Fixed `runOnFile` and `runOnList` running even when unset
- Various minor fixes related to compiler rewrite
- Various minor fixes related to EGP
### Removals
- Removed a clientside E2 file being added to the server erroneously
- Removed unused code in *file* extension
- Deprecated `clk`-type *file* functions
- Reduced network strings used for file operations

### Additions
- Added `iter` operator to replace `fea` operator
- Added `indexget` and `indexset` operators to replace `idx`
- Added `legacy` attribute for old format E2 functions
### Changes
- E2 functions now pass their evaluated arguments to the function
- Simplified E2 assignment and logical operators
### Removals
- Removed `fea`, `ass`, `inc`, `dec`, `and`, `or`, and `idx` operators

### Changes
- Minor optimization for entities with bone manipulations (*Community contribution*)
- Queue now tracks dupe codec revision to avoid using bad data
### Fixes
- Fixed `duplicator.RegisterEntityClass` with `"Data"` argument not being duped properly (*Community contribution*)

**Comment by thegrb93** [comment]
Aw they removed the ass operator...
